dm/master/simulator.go

```diff
@@ -0,0 +1,584 @@
+	"github.com/pingcap/parser/ast"
+	bf "github.com/pingcap/tidb-tools/pkg/binlog-filter"
+	"github.com/pingcap/tidb-tools/pkg/dbutil"
+	"github.com/pingcap/tidb-tools/pkg/filter"
+	router "github.com/pingcap/tidb-tools/pkg/table-router"
+	"go.uber.org/zap"
+)
+
+// SimulateTask does simulation on dm-master
+func (s *Server) SimulateTask(ctx context.Context, req *pb.SimulationRequest) (*pb.SimulationResponse, error) {
+	log.L().Info("", zap.Stringer("payload", req), zap.String("request", "FetchSourceInfo"))
+
+	cfg := config.NewTaskConfig()
+	err := cfg.Decode(req.Task)
+	if err != nil {
+		return &pb.SimulationResponse{
+			Result: false,
+			Msg:    errors.ErrorStack(err),
+		}, nil
+	}
+
+	_, stCfgs, err := s.generateSubTask(ctx, req.Task)
+	if err != nil {
+		return &pb.SimulationResponse{
+			Result: false,
+			Msg:    errors.ErrorStack(err),
+		}, nil
+	}
+
+	workerSet := make(map[string]struct{}, len(req.Workers))
+	if len(req.Workers) > 0 {
+		invalidWorkers := make([]string, 0, len(req.Workers))
+		for _, worker := range req.Workers {
+			if _, ok := s.workerClients[worker]; !ok {
+				invalidWorkers = append(invalidWorkers, worker)
+			} else {
+				workerSet[worker] = struct{}{}
+			}
+		}
+		if len(invalidWorkers) > 0 {
+			return &pb.SimulationResponse{
+				Result: false,
+				Msg:    fmt.Sprintf("%s relevant worker-client not found", strings.Join(invalidWorkers, ", ")),
+			}, nil
+		}
+	} else {
+		for worker := range s.workerClients {
+			workerSet[worker] = struct{}{}
+		}
+	}
+
+	var resp *pb.SimulationResponse
+	switch req.Op {
+	case pb.SimulateOp_TableRoute:
+		resp, err = simulateTableRoute(workerSet, req.TableList, s.cfg.DeployMap, stCfgs, cfg)
+		if err != nil {
+			return &pb.SimulationResponse{
+				Result: false,
+				Msg:    errors.ErrorStack(err),
+			}, nil
+		}
+	case pb.SimulateOp_BlackWhiteList:
+		resp, err = simulateBlackWhiteList(workerSet, req.TableList, s.cfg.DeployMap, stCfgs, cfg)
+		if err != nil {
+			return &pb.SimulationResponse{
+				Result: false,
+				Msg:    errors.ErrorStack(err),
+			}, nil
+		}
+	case pb.SimulateOp_EventFilter:
+		resp, err = simulateEventFilter(workerSet, req.Sql, s.cfg.DeployMap, stCfgs, cfg)
+		if err != nil {
+			return &pb.SimulationResponse{
+				Result: false,
+				Msg:    errors.ErrorStack(err),
+			}, nil
+		}
+	}
+	return resp, nil
+}
+
+func simulateTableRoute(workerSet map[string]struct{}, tables []string, deployMap map[string]string, stCfgs []*config.SubTaskConfig, cfg *config.TaskConfig) (*pb.SimulationResponse, error) {
+	simulationResultCap := len(deployMap)
+	if len(workerSet) > 0 {
+		simulationResultCap = len(workerSet)
+	}
+	simulationResults := make([]*pb.SimulationResult, 0, simulationResultCap)
+	resp := &pb.SimulationResponse{Result: true}
+
+	if len(tables) > 0 {
+		if len(workerSet) != 1 {
+			return nil, errors.Errorf("length of given workers should be 1, but is %d", len(workerSet))
+		}
+		var worker string
+		for worker = range workerSet {
+		}
```

Comment:
What is this...

Reply:
There supposed to be only one pair in workerSet and I want to extract the key from `workerSet`... Is there any better way?

Reply:
But it is actually an empty loop?

Reply:
OK, I got it...